### PR TITLE
add scripts, templates and settings for sending one-time domain invitations during data migration

### LINKED ISSUE #927: Data Migration: Domain Invitation Email send
### Issue Description

We will be sending emails separate of data ingestions. We want to avoid inviting already invited users while being able to update data from the to-end registry.
- For users with more than one domain, only send one email but include the links to all their domains not just the single domain.
- This script is a one time design for use with the migration. 
- This script should be able to be run multiple times with new sets of data, if the domain invitation has already been sent then it should NOT re-send for that email address 
  - The email is sent when transition_domain.email_sent=False. It should update this field in the table as it sends the emails. 
  - Ensure this is true even in the case where one user has multiple domains (all those domain rows should now have the same flag switched
- This should be VERY well tested. Include tests in the script or in a supporting script. Do not use it in production or with the actual emails from our host data until we are ready to execute the user Migration into our system. Test with fake emails or emails from team members.

### Additional Context (optional)

NOTE: Use the current domain invitation verbiage, run it by design to ensure it will be good given that it is for the migration vs normal invitations. 
https://docs.google.com/document/d/1UkQZdAmcGprXWdHX_PuT155CCNPbf_Gf/edit?pli=1#heading=h.iqv2v8lecjpn

retest this after #923 is done with all the data from verising
### Issue Links

blocked by: #923

### PR DESCRIPTION
**Pull request by dave-kennedy-ecs**
## Ticket

Resolves #927 

## Changes

- script generate_test_transition_domains.py added to generate test transition_domain data
- script send_domain_invitations.py added to send emails during migration
- added setting to settings.py to reference application's BASE_URL from scripts

## Context for reviewers

Separate ticket #1034 created to modify content of invitation emails.  So, email content does not need further review on this ticket.

generate_test_transition_domains.py is for testing purposes only; it is not used during migration; it is used to generate test transition_domains when they don't exist in order to test the script; this script takes -e option to add email addresses which will be assigned to the transition_domains
(ex. docker-compose exec app python manage.py generate_test_transition_domains -e email1,email2,email3)

send_domain_invitations.py is used to send domain invitations to the appropriate users, and mark the emails as sent in the transition_domain objects with email_sent=True; it can be run without actually sending emails and updating database by supplying no runtime argumens; to send emails and update database use the -s option
(ex. docker-compose exec app python manage.py send_domain_invitations -s)

## Setup

In your environment, run the following command to generate test transition_domains:
docker-compose exec app python manage.py generate_test_transition_domains -e your_email1,your_email2

You can then verify the transition domains in the admin interface in django.  Note that email_sent is unchecked for all transition_domains

Then run the script to send emails:
docker-compose exec app python manage.py send_domain_invitations -s

Check your email to verify that emails were received.
Check the admin interface in django and verify that email_send is checked for all transition_domains

Script can be repeated.  You can uncheck email_sent in admin interface and run again to resend emails.
You can also set the domain name in transition_domain to an invalid value to test error conditions when running script.

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [x] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [x] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

## Screenshots

(added 9/20/23)

Updated example below for single domain:

image

Updated example below for multiple domains:

image

### CODE REVIEW

File: src/registrar/management/commands/generate_test_transition_domains.py
Comment (zandercymatics) [inline src/registrar/management/commands/generate_test_transition_domains.py:35]: (Nice to have but optional) I'd add some error handling before this if options["emails"] is None, so we can throw a custom ValueError or something of that nature - "Parameter emails must be supplied" for example

File: src/registrar/management/commands/generate_test_transition_domains.py
Comment (zandercymatics) [inline src/registrar/management/commands/generate_test_transition_domains.py:25]: I like this approach

File: src/registrar/management/commands/generate_test_transition_domains.py
Comment (zandercymatics) [inline src/registrar/management/commands/generate_test_transition_domains.py:55]: I'd add some error handling in this function for test_emails (values of None for instance).

File: src/registrar/management/commands/generate_test_transition_domains.py
Comment (zandercymatics) [inline src/registrar/management/commands/generate_test_transition_domains.py:41]:
```suggestion
    def load_test_transition_domains(self, test_emails: list):
```
This doesn't have to be 'list' specifically if it breaks your function (just assuming the type), but type hinting here would be a good addition

File: src/registrar/management/commands/send_domain_invitations.py
Comment (zandercymatics) [inline src/registrar/management/commands/send_domain_invitations.py:62]: Given the context of this, I'd recommend changing these to be logger.info or (for the else) logger.error. I think this information makes sense to display to the end user

File: src/registrar/management/commands/send_domain_invitations.py
Comment (zandercymatics) [inline src/registrar/management/commands/send_domain_invitations.py:86]: This check seems redundant as you are checking for `not email_context["email"]` above... Your comment mentions checking for (not) the first item in the email array, did you mean to do that check here instead?

File: src/registrar/management/commands/send_domain_invitations.py
Comment (zandercymatics) [inline src/registrar/management/commands/send_domain_invitations.py:107]:
```suggestion
        # then append one more item
```
This seemed a bit unclear to me, I would recommend something like above

File: src/registrar/management/commands/send_domain_invitations.py
Comment (zandercymatics) [inline src/registrar/management/commands/send_domain_invitations.py:21]: This is a neat approach, I like this

File: src/registrar/management/commands/send_domain_invitations.py
Comment (zandercymatics) [inline src/registrar/management/commands/send_domain_invitations.py:95]: (Tangent/Unrelated) Not applicable here but when researching if we can grab the entire url within reverse, I found this which may be useful to you (hopefully) in future for django views if you do work on those: https://docs.djangoproject.com/en/2.0/ref/request-response/#django.http.HttpRequest.build_absolute_uri

File: src/registrar/management/commands/send_domain_invitations.py
Comment (zandercymatics) [inline src/registrar/management/commands/send_domain_invitations.py:112]: (Nice to have / Optional) I'd add an if check here for when emails_to_send is empty, something like `if len(emails_to_send) == 0: logger.info("No emails to send")`

File: src/registrar/management/commands/send_domain_invitations.py
Comment (zandercymatics) [inline src/registrar/management/commands/send_domain_invitations.py:125]: Ah, I see why you may be doing logger.debug in your handle function. Will this script always be restricted to debug only, or would you like to only restrict information but allow it to run otherwise?

File: src/registrar/templates/emails/transition_domain_invitation.txt
Comment (abroddrick) [inline src/registrar/templates/emails/transition_domain_invitation.txt:1]: @dave-kennedy-ecs I spoke with @Katherine-Osos  today and since @therealslimhsiehdy updated the domain_invitation text in [her newest pr](https://github.com/cisagov/getgov/pull/1026/files#diff-6837eab823a4ed83fd1e89c91d8a372ee3e5b8c8d8c2ba2109c8c8d565af3957), she would like to match the verbiage in there as much as possible. Mind updating this text to follow that PR accordingly?  Feel free though to wait until that PR is merged before updating this.

File: src/registrar/management/commands/generate_test_transition_domains.py
Comment (dave-kennedy-ecs) [reply src/registrar/management/commands/generate_test_transition_domains.py:35]: Thank you.  I considered this, but used required=True on the parser.add_argument, which essentially does the same thing.

File: src/registrar/management/commands/generate_test_transition_domains.py
Comment (dave-kennedy-ecs) [reply src/registrar/management/commands/generate_test_transition_domains.py:32]: Expanded the comment.  And handling logic in the parsing of command line arguments.  I do not think this script will be slated for reuse.

File: src/registrar/management/commands/generate_test_transition_domains.py
Comment (dave-kennedy-ecs) [reply src/registrar/management/commands/generate_test_transition_domains.py:41]: thanks

File: src/registrar/management/commands/send_domain_invitations.py
Comment (dave-kennedy-ecs) [reply src/registrar/management/commands/send_domain_invitations.py:86]: I intended this check to be here.  In this code, there is an ordered array of domains, sorted by username.  From that array, I create an array named emails_to_send of email_context to be sent, including data from one or more items from the first array in each item in the second array.  I have written checks that indicate when to add items to the second array.   The data to be added to the second array is stored in email_context while the domains array is looped through.

The first check indicates that the prior username in email_context does not match current username on the current item in the domains array.
The second check indicates that there is email_context to add to the second array.  This will not be the case when the first item in the domains array is processed.

File: src/registrar/management/commands/send_domain_invitations.py
Comment (dave-kennedy-ecs) [reply src/registrar/management/commands/send_domain_invitations.py:21]: thanks

File: src/registrar/management/commands/send_domain_invitations.py
Comment (dave-kennedy-ecs) [reply src/registrar/management/commands/send_domain_invitations.py:95]: Thank you.  Yes, I found this too, but request not available in this context.  So added to settings.py.

File: src/registrar/management/commands/send_domain_invitations.py
Comment (dave-kennedy-ecs) [reply src/registrar/management/commands/send_domain_invitations.py:112]: good suggestion.  i incorporated.

File: src/registrar/management/commands/send_domain_invitations.py
Comment (dave-kennedy-ecs) [reply src/registrar/management/commands/send_domain_invitations.py:125]: rethinking this.  i used debug intentionally so that this script could be run without the noise, and only errors would print.  but since the successful information is important to log for accountability, i have modified all of the debugs to infos.

File: src/registrar/management/commands/generate_test_transition_domains.py
Comment (dave-kennedy-ecs) [reply src/registrar/management/commands/generate_test_transition_domains.py:55]: i have now added the additional error handling to handle, and since redundant, did not add here as well.

File: src/registrar/templates/emails/transition_domain_invitation.txt
Comment (dave-kennedy-ecs) [reply src/registrar/templates/emails/transition_domain_invitation.txt:1]: @abroddrick @Katherine-Osos @therealslimhsiehdy I reviewed the pr for 482 yesterday, and I have updated the text.  I added some screenshots to the comment.

Note that this updated text for the email simplifies the code, as I no longer need to add individual urls for each domain to the email.  If you approve these updated screenshots, I will go back and simplify the code on this PR just a bit.

File: src/registrar/management/commands/send_domain_invitations.py
Comment (zandercymatics) [reply src/registrar/management/commands/send_domain_invitations.py:84]: Ah, yup, I see your or statement now. Thanks for pointing that out - I misread it as 'and' so I assumed it would never execute. That check makes sense to do, especially given the context you've given

**Review by zandercymatics — APPROVED** [review]
LGTM!

**Comment by abroddrick** [comment]
@zandercymatics did you run this locally and get it to work correctly? I don't see that checked off, and I want to make sure that was done here.

**Review by abroddrick — APPROVED** [review]
I will  approve as well.  These code comments are amazingly helpful and I want to say that your PR had one of the best descriptions of setup and what was done that I have seen yet. I really appreciate that as it made it much easier to review.

I know we are printing the domains and and emails when they fail, but at the end of the run could we add a print of the list of emails that triggered errors? I realize this may mean another list but it could be very helpful when trouble shooting in a production scenario.

**Comment by zandercymatics** [comment]
@abroddrick Ah my apologies! I did yesterday and it worked well, just spaced that portion

**Pull request merged**
